### PR TITLE
feat: custom challenges

### PR DESCRIPTION
**Pull request by mobitar**
Handles SNJS custom challenge support as implemented in https://github.com/standardnotes/snjs/pull/112.

### CODE REVIEW

File: app/assets/javascripts/views/challenge_modal/challenge_modal.ts [hunk 1]
```diff
@@ -118,34 +132,19 @@ class ChallengeModalCtrl extends PureViewCtrl {
...
   }
 
-  get title(): string {
+  get modalTitle(): string {
     if (this.challenge.reason === ChallengeReason.Migration) {
       return STRING_STORAGE_UPDATE;
     } else {
       return STRING_AUTHENTICATION_REQUIRED;
     }
   }
```
Comment (baptiste-grob) [inline app/assets/javascripts/views/challenge_modal/challenge_modal.ts:146]: Is the title not stored in the challenge now? If so we shouldn't use our own strings here

Comment (mobitar) [reply app/assets/javascripts/views/challenge_modal/challenge_modal.ts:146]: I believe this is the modal window title, which SNJS doesn't handle. Within the modal we have a prompt and subprompt, which is the title and subtitle in SNJS.

File: app/assets/javascripts/ui_models/application.ts [hunk 2]
```diff
@@ -167,21 +165,6 @@ export class WebApplication extends SNApplication {
     angular.element(document.body).append(el);
   }
 
-  async performProtocolUpgrade() {
```
Comment (mobitar) [inline app/assets/javascripts/ui_models/application.ts:170]: Handled directly by SNApplication now

File: app/assets/javascripts/views/application/application_view.ts [hunk 3]
```diff
@@ -258,24 +253,6 @@ class ApplicationViewCtrl extends PureViewCtrl {
     this.application!.componentManager!.presentPermissionsDialog = this.presentPermissionsDialog;
   }
 
-  showInvalidSessionAlert() {
```
Comment (mobitar) [inline app/assets/javascripts/views/application/application_view.ts:261]: Handled by SNJS directly

File: app/assets/templates/directives/account-menu.pug [hunk 4]
```diff
@@ -131,35 +130,10 @@
                   type='checkbox'
                   )
                 p.sk-p Merge local data ({{self.notesAndTagsCount()}} notes and tags)
-      .sk-panel-section(ng-if='self.state.formData.mfa')
```
Comment (mobitar) [inline app/assets/templates/directives/account-menu.pug:134]: MFA is now handled by SNJS directly via challenges

**Review by baptiste-grob — APPROVED** [review]
Very cool. Love to see web end up having less code :)
Please wait for the related SNJS PR to land before merging this one in.

**Pull request merged**